pkg/apis/stable/v1alpha1/gameserver.go

```diff
@@ -268,6 +259,45 @@ func (gs *GameServer) Validate() (bool, []metav1.StatusCause) {
 		}
 	}
 
+	return gs.validateConfig(causes)
+}
+
+// ValidateUpdate validates the GameServer configuration after the initial GameServer has been created.
+// If a GameServer is invalid there will be > 0 values in
+// the returned array
+func (gs *GameServer) ValidateUpdate() (bool, []metav1.StatusCause) {
+	var causes []metav1.StatusCause
+
+	// no host port when using dynamic PortPolicy only if we are still allocating a port.
+	if gs.Status.State == GameServerStatePortAllocation || gs.Status.State == "" {
```

Comment:
I think this works for 80% of cases, let me know if it doesn't - and also doesn't hardcode status values.

Switch the `ValidateUpdate` signature to match the other ones (also will need to validate against the old object, not the new one), and make the following change.

The TL;DR - you can only change ports, when the status.State is in flux. Otherwise you can't.
Here is a [patch.txt](https://github.com/GoogleCloudPlatform/agones/files/2830193/patch.txt), in case you want a quick way to apply the function (I've not done the webhook side or tested it)

There may be some bugs in here, but I hope you get the intent, and there is enough here to take it through, if the solution makes sense.

```go
// ValidateUpdate validates the GameServer configuration after the initial GameServer has been created.
// If a GameServer is invalid there will be > 0 values in
// the returned array
func (gs *GameServer) ValidateUpdate(new *GameServer) (bool, []metav1.StatusCause) {
	var causes []metav1.StatusCause

	// these values can never change
	if gs.Spec.Container != new.Spec.Container ||
		!equality.Semantic.DeepEqual(gs.Spec.Health, new.Spec.Health) ||
		gs.Spec.Scheduling != new.Spec.Scheduling ||
		!equality.Semantic.DeepEqual(gs.Spec.Template, new.Spec.Template) {
		causes = append(causes, metav1.StatusCause{
			Type:    metav1.CauseTypeFieldValueInvalid,
			Field:   "spec",
			Message: "spec cannot be updated",
		})
	}

	// port can be changed while status is in transition
	if gs.Status.State == new.Status.State &&
		len(causes) == 0 &&
		!equality.Semantic.DeepEqual(gs.Spec.Ports, new.Spec.Ports) {
		causes = append(causes, metav1.StatusCause{
			Type:    metav1.CauseTypeFieldValueInvalid,
			Field:   "spec",
			Message: "spec cannot be updated",
		})
	}

	return len(causes) == 0, causes
}
```

WDYT?